virtualmachine/aws/vm.go

```diff
@@ -358,6 +442,331 @@ func (vm *VM) Start() error {
+func (vm *VM) RevokeSecurityGroup() error {
+	svc, err := getService(vm.Region)
+	if err != nil {
+		return fmt.Errorf("Failed to get AWS service: %v", err)
+	}
+
+	secGrp := vm.SecurityGroups[0]
+
+	ec2IpPermissions := ToEc2IpPermissions(secGrp.IpPermissions)
+	input := &ec2.RevokeSecurityGroupIngressInput{
+		GroupId:       &secGrp.Id,
+		IpPermissions: ec2IpPermissions}
+	_, err = svc.RevokeSecurityGroupIngress(input)
+	if err != nil {
+		return fmt.Errorf("Failed to revoke security group ingress rules: %v", err)
+	}
+
+	ec2IpPermissionsEgress := ToEc2IpPermissions(
+		secGrp.IpPermissionsEgress)
+	egressInput := &ec2.RevokeSecurityGroupEgressInput{
+		GroupId:       &secGrp.Id,
+		IpPermissions: ec2IpPermissionsEgress}
+	_, err = svc.RevokeSecurityGroupEgress(egressInput)
+	if err != nil {
+		return fmt.Errorf("Failed to revoke security group egress rules: %v", err)
+	}
+
+	return nil
+}
+
+// CreateVolume: Creates a volume with given parameter
+func (vm *VM) CreateVolume() error {
+	svc, err := getService(vm.Region)
+	if err != nil {
+		return fmt.Errorf("Failed to get AWS service: %v", err)
+	}
+
+	volume := vm.Volumes[0]
+	instanceStatus, err := GetInstanceStatus(svc, vm.InstanceID)
+	if err != nil {
+		return fmt.Errorf("Failed to get availability zone of instance: %v", err)
+	}
+	volume.AvailabilityZone = instanceStatus.AvailabilityZone
+
+	input := GetVolumeInput(&volume)
+	response, err := svc.CreateVolume(input)
+	if err != nil {
+		return fmt.Errorf("Failed to create volume: %v", err)
+	}
+
+	if err := waitUntilVolumeState(svc, *response.VolumeId,
+		ec2.VolumeStateAvailable); err != nil {
+		return err
+	}
+
+	volume.VolumeId = *response.VolumeId
+	vm.Volumes[0] = volume
+
+	return nil
+}
+
+// AttachVolume: Attaches given volume to given instance
+func (vm *VM) AttachVolume() error {
+	svc, err := getService(vm.Region)
+	if err != nil {
+		return fmt.Errorf("Failed to get AWS service: %v", err)
+	}
+
+	volume := vm.Volumes[0]
+	input := &ec2.AttachVolumeInput{
+		Device:     &volume.DeviceName,
+		InstanceId: &vm.InstanceID,
+		VolumeId:   &volume.VolumeId}
+	_, err = svc.AttachVolume(input)
+	if err != nil {
+		return fmt.Errorf("Failed to attach volume (volumeId %s) to "+
+			"instance (instanceId %s): %v", volume.VolumeId, err)
+	}
+
+	if err := waitUntilVolumeState(svc, volume.VolumeId,
+		ec2.VolumeStateInUse); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DetachVolume: Detaches volume with given Id from instance
+func (vm *VM) DetachVolume() error {
+	svc, err := getService(vm.Region)
+	if err != nil {
+		return fmt.Errorf("Failed to get AWS service: %v", err)
+	}
+
+	volume := vm.Volumes[0]
+	input := &ec2.DetachVolumeInput{
+		VolumeId: &volume.VolumeId}
+	_, err = svc.DetachVolume(input)
+	if err != nil {
+		return fmt.Errorf("Failed to detach volume (volumeId %s) from "+
+			"instance (instanceId %s): %v", volume.VolumeId, err)
+	}
+
+	if err := waitUntilVolumeState(svc, volume.VolumeId,
```

Comment:
IMO, we should create a simple wrappers over waitUntilVolumeState and call them waitfordetach/ waitforattach etc and leave the actual state name dependency in the wait.go file itself.